File: src/main/java/io/vproxy/vfx/util/algebradata/PairData.java
package io.vproxy.vfx.util.algebradata;

public class PairData<A extends AlgebraData<A>, B extends AlgebraData<B>> implements AlgebraData<PairData<A, B>> {
    public final A a;
    public final B b;

    public PairData(A a, B b) {
        this.a = a;
        this.b = b;
    }

    @Override
    public PairData<A, B> plus(PairData<A, B> other) {
        return new PairData<>(a.plus(other.a), b.plus(other.b));
    }

    @Override
    public PairData<A, B> minus(PairData<A, B> other) {
        return new PairData<>(a.minus(other.a), b.minus(other.b));
    }

    @Override
    public PairData<A, B> multiply(double v) {
        return new PairData<>(a.multiply(v), b.multiply(v));
    }

    @Override
    public PairData<A, B> dividedBy(double v) {
        return new PairData<>(a.dividedBy(v), b.dividedBy(v));
    }

    @Override
    public String toString() {
        return "PairData(" + a + ", " + b + ')';
    }
}

File: src-ts/src/util/algebradata/PairData.ts
import { AlgebraData } from './AlgebraData.js';

export class PairData<
  A extends AlgebraData<A>,
  B extends AlgebraData<B>,
> implements AlgebraData<PairData<A, B>> {
  constructor(public readonly a: A, public readonly b: B) {}

  plus(o: PairData<A, B>): PairData<A, B> {
    return new PairData<A, B>(this.a.plus(o.a), this.b.plus(o.b));
  }
  minus(o: PairData<A, B>): PairData<A, B> {
    return new PairData<A, B>(this.a.minus(o.a), this.b.minus(o.b));
  }
  multiply(v: number): PairData<A, B> {
    return new PairData<A, B>(this.a.multiply(v), this.b.multiply(v));
  }
  dividedBy(v: number): PairData<A, B> {
    return new PairData<A, B>(this.a.dividedBy(v), this.b.dividedBy(v));
  }

  toString(): string { return `PairData(${this.a}, ${this.b})`; }
}
